Vac101-Merkle tree blog (#158)
vacp2p/vac.dev: rlog/2024-12-08-merkle-tree.mdx

The data representing the blockchain's state is compressed.
This compression addresses the issue of scale that would greatly restrict the pool of users.

Data compression alone is not the end goal.
As mentioned, it is essential for users to beable to validate the blockchain's history.
The property of compression and validation was solved in Bitcoin by the use of Merkle trees.
Merkle trees were first by Ralph Merkle in his dissertation [[1](https://www.ralphmerkle.com/papers/Thesis1979.pdf)].
A Merkle tree is a data structure that compresses a digest of data to a constant size while still providing a method for proving membership of elements of the digest.

## Tree structure

A tree is a special data structure that organizes nodes so that there is exactly one path between any two nodes.
The trees that we consider can be arranged in layers with multiple nodes (children) merge into a single node (parent) in the preceding node.
A single node exists in the base layer;
this special node is called the root node.
The highest level of the tree consists of parentless nodes called leaves.

![Image 1](/img/vac101/vac101_tree.png)

A binary tree has one additional property:

> each node has at most two children nodes.

Reviewer: ```suggestion
each node has exactly two children nodes.
```

If possible, this would compromise the ledger since the blockchain's history could be altered.
Fortunately, Merkle trees are quite secure.
In fact, Merkle trees can be used to both bind and hide a digest.

The Merkle tree is able to bind a digest with one of the properties of hash functions (see our previous Vac-101 [[2](https://vac.dev/rlog/vac101-fiat-shamir#hash-functions)] for information on hash functions).
A hash function is collision resistant; it is difficult for a malicious party to find two values share the same hash value.
This collision resistance property, essentially, fixes the input to each leaf and into their parent, their parent's parent, and so on.

In certain applications,
it may be desirable for the digest of a Merkle tree to be kept confidential.
This is achieved with the preimage resistant propery of hash functions.
A hash function is preimage resistant provided that it is difficult reverse the hashing operation. 
It would be necessary for a malicious party to find preimages to each node starting from the root node to determine original digest.

Now, we see that Merkle trees are secured structures that are tamper resistant.

> ### Proof of membership

Reviewer: I would clarify that the verification process needs the root hash to be known from a trusted source, if I'm not wrong.

The fingerprint of a piece of data $\ell$ indicates which leaf of the SMT it is contained in.
This means that a nonmembership of $\ell$ in the SMT becomes a matter of proving that $\mathsf{null}$ is contained in $\ell$'s location. 

It is crucial for the SMT to be sorted. 
Otherwise, an malicious party can append the entry $\ell$ to a random location.
This allows for the malicious party to contradictory proofs that prove both membership and nonmemberhip.
We note that the requirement that a SMT is sorted can be too strong of an assumption in centralized cases.
However, sorted is a necessary property of SMT for decentralized systems.

### Verkle Trees
A proof of membership grows in length as the Merkle tree grows.
The most obvious approach to remedy this scalability issue is to use Merkle trees where each node has more than 2 children.
However, as we saw earlier, this does not fix the issue.
An alternate approach is to use a different data structure: [Verkle trees](https://math.mit.edu/research/highschool/primes/materials/2018/Kuszmaul.pdf).
A Verkle tree replaces hash functions with polynomial commitments.
We will explore Verkle trees is a future Vac 101 edition.

## References

> - 1. [Secrey, Authentication, and Public Key Systems](https://www.ralphmerkle.com/papers/Thesis1979.pdf) 

Reviewer: ```suggestion
- 1. [Secrecy, Authentication, and Public Key Systems](https://www.ralphmerkle.com/papers/Thesis1979.pdf) 
```

---
title: 'Vac 101 - Climbing Merkle Trees'
date: 2024-12-08 12:00:00
authors: marvin
published: true
slug: climbing-merkle-trees
categories: research

toc_min_heading_level: 2
toc_max_heading_level: 4
---

In this post, we introduce a crucial data structure used throughout web3.

<!--truncate-->

## Introduction

A large amount of data is swapped between users on a blockchain in the form of transactions.
Over the entire life of a blockchain,
the storage space required to maintain a copy of every transaction becomes untenable for most users.
However, the integrity of a blockchain relies on a large pool of users that can validate the blockchain's history from its inception to its present state.
The data representing the blockchain's state is compressed.
This compression addresses the issue of scale that would greatly restrict the pool of users.

Data compression alone is not the end goal.

> As mentioned, it is essential for users to beable to validate the blockchain's history.

Reviewer 2: ```suggestion
As mentioned, it is essential for users to be able to validate the blockchain's history.
```

> The property of compression and validation was solved in Bitcoin by the use of Merkle trees.

Reviewer 2: We can mention some use case examples, such as nullifying, proving membership (e.g. RLN), lightclients.

## Merkle trees

A Merkle tree is a specialized tree in which each of its nodes contain the evaluation of a hash function.
Merkle trees are usually taken to have a binary tree structure.
As such, the presentation we provide in this section will be for binary trees.

### Construction
In this section we show how Merkle tree are constructed to compress a digest $D$.
Suppose that the digest $D$ consists of $2^n$ entries;
we assume that the digest $D$ has this many entries since a Merkle tree is a binary tree.
Additionally, each digest can be padded to ensure that $D$ has the desired number of entries.

Each leaf of the Merkle tree contains the hash of a digest entry.
Each parent node contains the hash of the concatenation of their children nodes.
Through this iterative construction, we reach the root of the tree.
The value contained in the root node is called the root hash.
The root hash is a compressed representation of the digest $D$.

![Image 2](/img/vac101/vac101_merkle_tree.png)

Each node in the Merkle tree is computed by taking a hash.
Since a binary tree with $2^n$ leaves has $2^{n+1}-1$ nodes,
then we need to evaluate $2^{n+1}-1$ hashes to construct the Merkle tree.

> ### Merkle tree intregity

Reviewer 2: ```suggestion
### Merkle tree intregrity
```

A large quantity of data can be compressed to a single hash value.
A natural question to ask is: could a clever party find another digest that yields a Merkle tree with the same root hash?
If possible, this would compromise the ledger since the blockchain's history could be altered.
Fortunately, Merkle trees are quite secure.
In fact, Merkle trees can be used to both bind and hide a digest.

The Merkle tree is able to bind a digest with one of the properties of hash functions (see our previous Vac-101 [[2](https://vac.dev/rlog/vac101-fiat-shamir#hash-functions)] for information on hash functions).

> A hash function is collision resistant; it is difficult for a malicious party to find two values share the same hash value.

Reviewer 2: ```suggestion
A hash function is collision resistant; it is infeasible for a malicious party to find two values share the same hash value.
```

This collision resistance property, essentially, fixes the input to each leaf and into their parent, their parent's parent, and so on.

In certain applications,
it may be desirable for the digest of a Merkle tree to be kept confidential.
This is achieved with the preimage resistant propery of hash functions.
A hash function is preimage resistant provided that it is difficult reverse the hashing operation. 
It would be necessary for a malicious party to find preimages to each node starting from the root node to determine original digest.

Now, we see that Merkle trees are secured structures that are tamper resistant.

### Proof of membership
An interesting and critical use of Merkle trees is its ability to prove any piece of data is a part of its digest.

> This can be done with logarthmic storage and logartihmic computation time.

Reviewer 2: ```suggestion
This can be done with logarthmic storage and logarithmic computation time.
```

Suppose that we want to show that data $\ell$ is part of the Merkle tree's digest.
Additionally, suppose that $\mathsf{hash}$ is the hash function used to construct the tree.
We assume that the hash function $\mathsf{hash}$ can be computed in constant-time for any input.

> Suppose that the Prover provides data $\ell$ to the Verifier,

Reviewer 2: ```suggestion
Suppose that a Prover provides data $\ell$ to a Verifier,
```

The leaf index $i$ indicates whether a hash value provided by the Prover is a left or right sibling.
This is done by looking at the binary expansion of $i$.

The Verifier can compute the leaf $h_0 = \mathsf{hash}(\ell)$.
Next, using $h_0$'s sibling, $h'_0$, provided by the Prover:
the Verifier can compute $h_1 = \mathsf{hash}(h_0 \|h'_0)$ or $h_1 = \mathsf{hash}(h'_0 \| h_0)$
depending on whether $h'_0$ is a left or right sibling.
This pathing continues until the Verifier either successfully computes the root hash (in $n+1$ hashes) or fails to.

The Prover has to provide $n$ sibling nodes for the proof of membership.

There is a key detail that is essential for the proof of membership to be secure.
The root hash has to be provided to the Verifier prior to the selection of data $\ell$.
Otherwise, Prover could generate a series of hash values (with the corresponding root hash) to forge a proof of membership.

## Extensions of Merkle trees

Merkle trees can be extended in multiple ways.
In this section, we explore a select few of these extensions.

> ### Sparse Merkle trees

Reviewer 2: adding some white-listed implementations can be good

Author: Added a repo for Verkle tree. Cannot locate a white-listed implementation of sparse merkle trees.

A simple method for this is to use a hash function to provide fingerprints to the data.
Each hash using either SHA-256 or Keccak has 256-bits.
Our entire digest could consist of a maximum of $2^{256}$ entries.
This assumes that our digest does not contain collisions.

The fingerprint of a piece of data $\ell$ indicates which leaf of the SMT it is contained in.
This means that a nonmembership of $\ell$ in the SMT becomes a matter of proving that $\mathsf{null}$ is contained in $\ell$'s location. 

It is crucial for the SMT to be sorted. 
Otherwise, an malicious party can append the entry $\ell$ to a random location.
This allows for the malicious party to contradictory proofs that prove both membership and nonmemberhip.
We note that the requirement that a SMT is sorted can be too strong of an assumption in centralized cases.
However, sorted is a necessary property of SMT for decentralized systems.

### Verkle Trees
A proof of membership grows in length as the Merkle tree grows.
The most obvious approach to remedy this scalability issue is to use Merkle trees where each node has more than 2 children.

> However, as we saw earlier, this does not fix the issue.

Reviewer 2: Which line we saw this does not fix the issue?

Reviewer 2: We can also mendtion k-ary merkle trees and MPT for further readings.

Author: Added some details about $k$-nary Merkle tree proof size here. Will add references for MPT and $k$-nary.

An alternate approach is to use a different data structure: [Verkle trees](https://math.mit.edu/research/highschool/primes/materials/2018/Kuszmaul.pdf).

> A Verkle tree replaces hash functions with polynomial commitments.

Reviewer 2: ref for the PCS can be good

Author: Added references
